iterators.rb

```diff
@@ -0,0 +1,132 @@
+# rubocop:disable  Metrics/ModuleLength, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+module Enumerable
+  def my_each
+    index = 0
+    while index < size
+
+      case self.class.name
+      when 'Hash' then yield(keys[index], self[keys[index]])
+      when 'Array' then yield(self[index])
+      when 'Range' then yield(to_a[index])
+      end
+      index += 1
+    end
+  end
+
+  def my_each_with_index
+    index = 0
+    while index < size
+      yield(self[index], index)
+      index += 1
+    end
+  end
+
+  def my_select
+    return enum_for unless block_given?
+
+    array = []
+    my_each do |element|
+      array.push(element) if yield(element) == true
+    end
+    array
+  end
+  alias my_find_all my_select
+  alias my_filter my_select
+
+  def my_all?(pattern = nil)
+    is_equal = true
+    if pattern.nil?
+      if block_given?
+        my_each { |val| is_equal = false unless yield(val) or size.zero? }
+      else my_each { |val| is_equal = false unless val == true }
+      end
+    elsif pattern.is_a?(Module)
+      if pattern.is_a?(Regexp)
+        my_each { |val| is_equal = false unless pattern.match(val) }
+      else
+        my_each { |val| is_equal = false unless val.is_a?(pattern) }
+      end
+    end
+    is_equal
+  end
+
+  def my_any?
+    my_each do |val|
+      return true if yield(val)
+    end
+    false
+  end
+
+  def my_none?(pattern = nil)
+    is_true = true
+    if pattern.nil?
+      if block_given?
+        my_each { |val| is_true = false if yield(val) or size.zero? }
+      else my_each { |val| is_true = false if val == true }
+      end
+    elsif pattern.is_a?(Module)
+      if pattern.is_a?(Regexp)
+        my_each { |val| is_true = false if pattern.match(val) }
+      else
+        my_each { |val| is_true = false if val.is_a?(pattern) }
+      end
+    end
+    is_true
+  end
+
+  def my_count(args = 0)
+    if block_given?
+      counter = 0
+      my_each do |element|
+        counter += 1 if yield(element)
+      end
+      counter
+    elsif args.positive?
+      filtered = my_select { |element| element == args }
+      filtered.size
+    else
+      size
+    end
+  end
```

Comment:
* [ ] should have working `my_count` identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_count

- [ ] my_count counts the number of items in enum that are equal to the item if an argument is given
```ruby
[5, 4, 1, 4, 5, 0].my_count(0) == [5, 4, 1, 4, 5, 0].count(0) => 1
```
- Yours is giving the wrong answer
![image](https://user-images.githubusercontent.com/52209067/80292959-2cdf0680-8718-11ea-9735-d15e39d02681.png)

- Real count
![image](https://user-images.githubusercontent.com/52209067/80292969-4c762f00-8718-11ea-8228-3b8f9925eeaa.png)

Reply:
**Fixed**